File: java/src/algorithms/eratosthenes_sieve/EratosthenesSieve.java
package algorithms.eratosthenes_sieve;

import java.util.ArrayList;
import java.util.BitSet;
import java.util.List;

/**
 * Implementation of the sieve of Eratosthenes algorithm for finding
 * all prime numbers up to a specified limit.
 *
 * @author Sebastian Claici
 */
public class EratosthenesSieve {

    /**
     * Returns a list of all prime numbers up to (and including) N.
     *
     * Time complexity: O(N * log(log(N)))
     *
     * @param N Upper bound up to which primes should be computed.
     * @return A list of integers representing the prime numbers up to N.
     */
    public static List<Integer> sieve(int N) {
        List<Integer> primes = new ArrayList<Integer>();

        // Initially all numbers are considered prime
        BitSet bitSet = new BitSet(N);
        bitSet.flip(0, bitSet.size());

        for (int i = 2; i <= N; ++i) {
            if (bitSet.get(i)) {
                primes.add(i);

                long limit = (long) i * i;
                if (limit > N)
                    continue;

                // Mark all multiples of i as non-prime
                for (int j = i * i; j <= N; j += i)
                    bitSet.set(j, false);
            }
        }

        return primes;
    }

    // Test code
    public static void main(String[] args) {
        List<Integer> primes = sieve(100);
        int[] primesTo100 = {2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53,
                59, 61, 67, 71, 73, 79, 83, 89, 97};

        boolean correct = true;
        for (int i = 0; i < primes.size(); ++i)
            if (primes.get(i) != primesTo100[i]) {
                correct = false;
                break;
            }

        if (!correct)
            System.out.println("Wrong answer");
        else
            System.out.println("Correct");
    }
}

File: cpp/algorithms/eratosthenes_sieve/eratosthenes_sieve.cpp
#include <fstream>
#include <vector>

// Returns a list of all prime numbers up to N (inclusive). Uses the
// sieve of Eratosthenes technique to filter out numbers that are not
// prime.
//
// Time complexity: O(N * log(log(N)))
//
// Args:
//      N: number up to which prime numbers should be recovered
//
// Returns: a vector containing all prime numbers up to N in sorted
//      order
std::vector<int> sieve(const int N)
{
    std::vector<int> primes;

    // initially, all numbers are considered prime
    std::vector<bool> is_prime(N, true);

    for (int i = 2; i <= N; ++i) {
        if (is_prime[i]) {
            primes.push_back(i);
            
            auto limit = static_cast<long long>(i) * i;
            if (limit > N)
                continue;
            // mark all multiples of i as non-prime
            for (int j = i * i; j <= N; j += i)
                is_prime[j] = false;
        }
    }

    return primes;
}

// Test code
int main()
{
    std::ifstream fin("ciur.in");
    std::ofstream fout("ciur.out");

    int N;
    fin >> N;

    auto primes = sieve(N);
    fout << primes.size() << "\n";

    return 0;
}
